interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```
@@ -0,0 +1,319 @@
+    clientQueueSize += delta;
+    if (clientQueueSize < queueLength) {
+      notifyAll();
+    }
+  }
+
+  private synchronized int getServerChunk() {
+    while (serverQueueSize == queueLength) {
+      try {
+        wait();
+      } catch (InterruptedException e) {
+        return 0;
+      }
+    }
+    return Math.max(0, Math.min(chunkSize, queueLength - serverQueueSize));
+  }
+
+  private synchronized int getClientChunk() {
+    while (clientQueueSize == queueLength) {
+      try {
+        wait();
+      } catch (InterruptedException e) {
+        return 0;
+      }
+    }
+    return Math.max(0, Math.min(chunkSize, queueLength - clientQueueSize));
+  }
+
+  public static void main(String[] args) throws IOException {
+    TrafficControlProxy p = new TrafficControlProxy();
+    p.start();
+  }
+
+  private final class ClientReader implements Runnable {
+
+    private final DataInputStream clientIn;
+
+    ClientReader(DataInputStream clientIn) {
+      this.clientIn = checkNotNull(clientIn);
+    }
+
+    @Override
+    public void run() {
+      while (!shutDown) {
+        try {
+          int nextChunk = getClientChunk();
+          if (nextChunk != 0) {
+            byte[] request = new byte[nextChunk];
+            int readableRequestBytes = clientIn.read(request);
+            long sendTime = System.nanoTime() + latency;
+            clientQueue.offer(new Message(sendTime, request, readableRequestBytes));
+            incrementClientQueueSize(readableRequestBytes);
+          }
+        } catch (IOException e) {
+          shutDown = true;
+        }
+      }
+    }
+  }
+
+  private final class ServerReader implements Runnable {
+
+    private final DataInputStream serverIn;
+
+    ServerReader(DataInputStream serverIn) {
+      this.serverIn = checkNotNull(serverIn);
+    }
+
+    @Override
+    public void run() {
+      while (!shutDown) {
+        try {
+          int nextChunk = getServerChunk();
+          if (nextChunk != 0) {
```

Why not just let the exception propagate?   You always interrupt when you are done with work, so there isn't any chance of recovery.